### PR TITLE
getCoinDistributionsForReview implementation

### PR DESCRIPTION
**Pull request by ice-cronus**
API to get distributions pending review by batches

### CODE REVIEW

File: cmd/freezer-refrigerant/coin-distribution.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+// GetCoinDistributionsForReview godoc
+//
+//	@Schemes
+//	@Description	Fetches data of pending coin distributions for review.
+//	@Tags			CoinDistribution
+//	@Accept			json
+//	@Produce		json
+//	@Param			Authorization	header		string								true	"Insert your access token"	default(Bearer <Add access token here>)
+//	@Param			x_client_type	query		string								false	"the type of the client calling this API. I.E. `web`"
+//	@Param			request			body		GetCoinDistributionForReviewParams	true	"Request params"
```
Comment (ice-ares) [inline cmd/freezer-refrigerant/coin-distribution.go]: No body. Everything in query

File: cmd/freezer-refrigerant/coin-distribution.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+) (*server.Response[CoinDistributionsForReview], *server.Response[server.ErrorResponse]) {
+	if req.AuthenticatedUser.Role != adminRole {
+		return nil, server.Forbidden(errors.Errorf("insufficient role: %v, admin role required", req.AuthenticatedUser.Role))
+	}
+	if req.Data.Limit == 0 {
+		req.Data.Limit = defaultDistributionLimit
+	}
+	cursor, distributions, err := s.coinDistributionRepository.GetCoinDistributionsForReview(ctx, req.Data.Cursor, req.Data.Limit)
+	if err != nil {
+		return nil, server.Unexpected(err)
```
Comment (ice-ares) [inline cmd/freezer-refrigerant/coin-distribution.go]: Wrap it

File: cmd/freezer-refrigerant/contract.go [hunk 3]
```diff
@@ -30,13 +31,25 @@ type (
...
 		Allocation uint8  `json:"allocation" required:"true" maximum:"100" example:"100"`
 	}
+
+	CoinDistributionsForReview struct {
+		Distributions []*coindistribution.CoinDistibutionForReview `json:"distributions"`
+		Cursor        uint64                                       `json:"cursor" example: 5065`
+	}
+
+	GetCoinDistributionForReviewParams struct {
+		Cursor uint64 `json:"cursor" example:"5065"`
```
Comment (ice-ares) [inline cmd/freezer-refrigerant/contract.go]: move both to form

File: cmd/freezer-refrigerant/contract.go [hunk 4]
```diff
@@ -50,12 +63,15 @@ const (
 	miningDisabledErrorCode                                  = "MINING_DISABLED"
 	noExtraBonusAvailableErrorCode                           = "NO_EXTRA_BONUS_AVAILABLE"
 	extraBonusAlreadyClaimedErrorCode                        = "EXTRA_BONUS_ALREADY_CLAIMED"
+
+	defaultDistributionLimit = 5000
```
Comment (ice-ares) [inline cmd/freezer-refrigerant/contract.go:67]: Is that the optimal batch size?

File: cmd/freezer-refrigerant/freezer_refrigerant.go [hunk 5]
```diff
@@ -36,10 +37,12 @@ func main() {
 
 func (s *service) RegisterRoutes(router *server.Router) {
 	s.setupTokenomicsRoutes(router)
+	s.setupCoinDistributionRoutesRoutes(router)
 }
 
 func (s *service) Init(ctx context.Context, cancel context.CancelFunc) {
 	s.tokenomicsProcessor = tokenomics.StartProcessor(ctx, cancel)
+	s.coinDistributionRepository = coindistribution.NewRepository(ctx, cancel)
```
Comment (ice-ares) [inline cmd/freezer-refrigerant/freezer_refrigerant.go:46]: U're not closing it

Comment (ice-ares) [reply cmd/freezer-refrigerant/freezer_refrigerant.go:46]: Also, check health for it

File: coin-distribution/contract.go [hunk 6]
```diff
@@ -19,6 +20,19 @@ type (
...
+	Repository interface {
+		GetCoinDistributionsForReview(ctx context.Context, cursor, limit uint64) (updatedCursor uint64, distributions []*CoinDistibutionForReview, err error)
+	}
+
+	CoinDistibutionForReview struct {
+		CreatedAt          *time.Time `json:"time" swaggertype:"string" example:"2022-01-03T16:20:52.156534Z"`
+		Iceflakes          string     `json:"iceflakes" swaggertype:"string" example:"100000000000000"`
+		Username           string     `json:"username" swaggertype:"string" example:"myusername"`
+		ReferredByUsername string     `json:"referredByUsername" swaggertype:"string" example:"myrefusername"`
+		UserID             string     `json:"userID" swaggertype:"string" example:"12746386-03de-44d7-91c7-856fa66b6ed6"`
```
Comment (ice-ares) [inline coin-distribution/contract.go]: json userId

File: coin-distribution/distribution_review.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/ice-blockchain/wintr/connectors/storage/v2"
+)
+
+func (r *repository) GetCoinDistributionsForReview(ctx context.Context, cursor, limit uint64) (updCursor uint64, distributions []*CoinDistibutionForReview, err error) {
+	sql := `SELECT * FROM coin_distributions_pending_review WHERE internal_id > $1 ORDER BY internal_id LIMIT $2`
```
Comment (ice-ares) [inline coin-distribution/distribution_review.go:14]: Add index on that then

**Review by ice-ares — CHANGES_REQUESTED** [review]

File: cmd/freezer-refrigerant/contract.go (hunk 4 above)
Comment (ice-cronus) [reply cmd/freezer-refrigerant/contract.go:67]: I tried in full local setup with 1k, 5k and 10k, and got best results for 5k (like 48-50s for 2M in total, 130-150ms per batch), on real deployment network latency will affect also, I'll retry it on staging once we deploy.

Comment (ice-cronus) [reply cmd/freezer-refrigerant/contract.go:67]: Well with index on that field it is 14s for 2M

File: cmd/freezer-refrigerant/coin-distribution.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/ice-blockchain/wintr/server"
+)
+
+func (s *service) setupCoinDistributionRoutesRoutes(router *server.Router) {
```
Comment (ice-ares) [inline cmd/freezer-refrigerant/coin-distribution.go]: setupCoinDistributionRoutes

File: cmd/freezer-refrigerant/coin-distribution.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+// GetCoinDistributionsForReview godoc
+//
+//	@Schemes
+//	@Description	Fetches data of pending coin distributions for review.
+//	@Tags			CoinDistribution
+//	@Accept			json
+//	@Produce		json
+//	@Param			Authorization	header		string	true	"Insert your access token"	default(Bearer <Add access token here>)
+//	@Param			x_client_type	query		string	false	"the type of the client calling this API. I.E. `web`"
```
Comment (ice-ares) [inline cmd/freezer-refrigerant/coin-distribution.go:27]: run make generate to format it

File: coin-distribution/coin_distribution.go [hunk 10]
```diff
@@ -39,6 +39,12 @@ func MustStartCoinDistribution(ctx context.Context, cancel context.CancelFunc) C
 	return cd
 }
 
+func NewRepository(ctx context.Context, cancel context.CancelFunc) Repository {
+	repo := &repository{db: storage.MustConnect(context.Background(), ddl, applicationYamlKey)}
```
Comment (ice-ares) [inline coin-distribution/coin_distribution.go]: ctx

Comment (ice-ares) [reply coin-distribution/coin_distribution.go]: _ context.CancelFunc

**Pull request merged**